### PR TITLE
Full Unicode support for TextBlock

### LINKED ISSUE #2548: Font style "Underline"
How can I set the font style "Underline"?

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?
This PR introduces multiple things

**TextFormatter**
The TextFormatter will be responsible for doing all the heavy lifting for all the text processing that is needed to build up a text layout. It functions as a service to implement your own layouts and it makes it possible to implement text editors etc.

**TextBlock**
The TextBlock control currently uses a FormattedText implementation that doesn't work well with advanced Unicode scenarios. This PR replaces FormattedText with the TextLayout class that is able to deal with all kinds of Unicode texts and produces expected results.

## Fixes
#2548
#3522
#1817

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

File: tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs
Comment (grokys) [inline tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs]: Why were tests tests moved from UnitTests to RenderTests? As far as I can see they're actually unit tests, not render tests (they're not rendering to a bitmap etc).

File: src/Avalonia.Visuals/Media/Text/TextLayout.cs
Comment (grokys) [inline src/Avalonia.Visuals/Media/Text/TextLayout.cs]: Commented-out code.

File: src/Avalonia.Controls/Primitives/AccessText.cs
Comment (grokys) [inline src/Avalonia.Controls/Primitives/AccessText.cs]: I don't think `AccessText` should inherit from `TextPresenter`. `TextPresenter` is meant to display the contents of a `TextBox`; `AccessText` isn't a text box (doesn't accept input), it's just a `TextBlock` that displays an underline when `ShowAccessKey` is set.

File: src/Avalonia.Visuals/Media/Text/TextLayout.cs
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/Text/TextLayout.cs]: Remove this

File: src/Avalonia.Visuals/Media/Text/TextRun.cs
Comment (Gillibald) [inline src/Avalonia.Visuals/Media/Text/TextRun.cs]: Wrong name

File: tests/Avalonia.RenderTests/Media/VisualBrushTests.cs
Comment (grokys) [inline tests/Avalonia.RenderTests/Media/VisualBrushTests.cs]: What is the problem with `VisualBrush` and text rendering?

**Comment by grokys** [comment]
In addition, it would be nice to get some class-level doc comments for the public classes in `Avalonia.Visuals\Media\Text` to help people not familiar with text layout internals.

File: tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs
Comment (Gillibald) [reply tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs]: `AssetLoader` is also defined in `Avalonia.UnitTests` but `Avalonia.Skia.UnitTests` requires this reference. So I had to move the tests to `RenderTests`. First I tried to get around this by explicitly activateing the right type by using `Activator.CreateInstance`. but it caused issues at other places.

File: tests/Avalonia.RenderTests/Media/VisualBrushTests.cs
Comment (Gillibald) [reply tests/Avalonia.RenderTests/Media/VisualBrushTests.cs]: I think tiling didn't work well. Will run the test again and attach an image.

File: src/Avalonia.Controls/Primitives/AccessText.cs
Comment (Gillibald) [reply src/Avalonia.Controls/Primitives/AccessText.cs]: Will revert the change. Makes sense👍

File: tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs
Comment (grokys) [reply tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs]: > AssetLoader is also defined in Avalonia.UnitTests but Avalonia.Skia.UnitTests requires this reference. 

Sorry, I don't understand - requires which reference?

Comment (Gillibald) [reply tests/Avalonia.Direct2D1.RenderTests/FontManagerImplTests.cs]: `Avalonia.Skia.UnitTests' has a project reference to `Avalonia.UnitTests`

`AssetLoader` is defined in `Avalonia.Shared.PlatformSupport` and `Avalonia.UnitTests`

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs]: Not directly related to this PR: We already have an issue for spawning dedicated timers for handling undo in `TextBox` and this makes me wonder if we also really need a dedicated timer to just blink the caret. Makes it fairly expensive to construct these controls.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs]: Nit: Don't we usually use expression bodied properties in new code?

File: src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs
Comment (MarchingCube) [inline src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs]: I know this is not your code, but this code won't work correctly on an actual big endian machine (even though it may never run on such). It should only reverse when on little endian machine.

File: src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs
Comment (KvanTTT) [inline src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs]: Also, maybe it makes sense to get rid of allocations?

```cs
public static uint ReadUInt32BE(this BinaryReader reader)
{
    var byte0 = reader.ReadByte();
    var byte1 = reader.ReadByte();
    var byte2 = reader.ReadByte();
    var byte3 = reader.ReadByte();
    return ((uint) byte3 << 24) | ((uint) byte2 << 16) | ((uint) byte1 << 8) | byte0;
}
```

File: src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs
Comment (KvanTTT) [inline src/Avalonia.Visuals/Media/TextFormatting/Unicode/BinaryReaderExtensions.cs]:
```cs
public static void WriteBE(this BinaryWriter writer, int value)
{
    writer.Write((byte)(value >> 24));
    writer.Write((byte)(value >> 16));
    writer.Write((byte)(value >> 8));
    writer.Write((byte)value);
}
```

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (grokys) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: Yeah, lets get this in first and then we can think about optimizing.

**Comment by Gillibald** [comment]
I can't guarantee that there will be no breaking changes in the feature. Because the current version of the TextFormatter can only deal with text and not with Controls. So at some point in the future, the API surface will change.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (Gillibald) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: This is just a copy paste because TextPresenter no longer inherits from TextBlock. If you insist on this change I am fine with changing to expression bodies.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (Gillibald) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: New implementation will have a dedicated control for displaying a caret. The shape of the caret depends on the current font etc. So it makes sense to separate this logic.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs
Comment (MarchingCube) [reply src/Avalonia.Controls/Presenters/TextPresenter.cs]: I don’t really mind, but I usually convert to expression bodies. This is just a nitpick so ultimately it is up to you.

**Comment by Gillibald** [comment]
Just realized that I have to implement https://unicode.org/reports/tr29/ before this can be merged. Shouldn't be too complicated.

**Comment by Gillibald** [comment]
WIP again. Have to polish things.

**Comment by Gillibald** [comment]
Making this WIP again because I want to introduce a text source first to avoid future breaking changes.

**Comment by Gillibald** [comment]
This is now ready for inlines support

**Review by grokys — APPROVED** [review]
Again, I can't say I'm really qualified to give this the review it deserves, but what I do understand of it looks excellent.

**Comment by grokys** [comment]
@Gillibald i don't seem to be able to update this branch to master, could you make sure the upstream repo has write access?

**Comment by Gillibald** [comment]
Somehow the right wasn't granted. Have changed that.

**Pull request merged**
